internal/gps/pkgtree/digest.go

```diff
@@ -0,0 +1,442 @@
+
+	// when walking vendor root hierarchy, ignore file system nodes of the
+	// following types.
+	skipSpecialNodes = os.ModeDevice | os.ModeNamedPipe | os.ModeSocket | os.ModeCharDevice
+)
+
+// DigestFromPathname returns a deterministic hash of the specified file system
+// node, performing a breadth-first traversal of directories. While the
+// specified prefix is joined with the pathname to walk the file system, the
+// prefix string is eliminated from the pathname of the nodes encounted when
+// hashing the pathnames, so that the resultant hash is agnostic to the absolute
+// root directory path of the nodes being checked.
+//
+// This function ignores any file system node named `vendor`, `.bzr`, `.git`,
+// `.hg`, and `.svn`, as these are typically used as Version Control System
+// (VCS) directories.
+//
+// Other than the `vendor` and VCS directories mentioned above, the calculated
+// hash includes the pathname to every discovered file system node, whether it
+// is an empty directory, a non-empty directory, empty file, non-empty file, or
+// symbolic link. If a symbolic link, the referent name is included. If a
+// non-empty file, the file's contents are incuded. If a non-empty directory,
+// the contents of the directory are included.
+//
+// While filepath.Walk could have been used, that standard library function
+// skips symbolic links, and for now, we want the hash to include the symbolic
+// link referents.
+func DigestFromPathname(prefix, pathname string) ([]byte, error) {
+	// Create a single hash instance for the entire operation, rather than a new
+	// hash for each node we encounter.
+	h := sha256.New()
+
+	// Initialize a work queue with the os-agnostic cleaned up pathname. Note
+	// that we use `filepath.Clean` rather than `filepath.Abs`, because the hash
+	// has pathnames which are relative to prefix, and there is no reason to
+	// convert to absolute pathname for every invocation of this function.
+	prefix = filepath.Clean(prefix) + pathSeparator
+	prefixLength := len(prefix) // store length to trim off pathnames later
+	pathnameQueue := []string{filepath.Join(prefix, pathname)}
+
+	var written int64
+
+	// As we enumerate over the queue and encounter a directory, its children
+	// will be added to the work queue.
+	for len(pathnameQueue) > 0 {
+		// Unshift a pathname from the queue (breadth-first traversal of
+		// hierarchy)
+		pathname, pathnameQueue = pathnameQueue[0], pathnameQueue[1:]
+
+		fi, err := os.Lstat(pathname)
+		if err != nil {
+			return nil, errors.Wrap(err, "cannot Lstat")
+		}
+		mode := fi.Mode()
+
+		// Skip file system nodes we are not concerned with
+		if mode&skipSpecialNodes != 0 {
+			continue
+		}
+
+		// Write the prefix-stripped pathname to hash because the hash is as
+		// much a function of the relative names of the files and directories as
+		// it is their contents. Added benefit is that even empty directories
+		// and symbolic links will effect final hash value. Use
```

Comment:
Never a nit too small. I just wish my brain's English linter threw errors when I made this sort of typo... ;)